packages/editor-ui/src/stores/collaboration.store.ts

```diff
@@ -1,86 +1,118 @@
 
 /**
  * Store for tracking active users for workflows. I.e. to show
  * who is collaboratively viewing/editing the workflow at the same time.
  */
 export const useCollaborationStore = defineStore(STORES.COLLABORATION, () => {
 	const pushStore = usePushConnectionStore();
-	const workflowStore = useWorkflowsStore();
+	const workflowsStore = useWorkflowsStore();
 	const usersStore = useUsersStore();
+	const uiStore = useUIStore();
 
-	const usersForWorkflows = ref<ActiveUsersForWorkflows>({});
-	const pushStoreEventListenerRemovalFn = ref<(() => void) | null>(null);
+	const route = useRoute();
+	const { addBeforeUnloadEventBindings, removeBeforeUnloadEventBindings, addBeforeUnloadHandler } =
+		useBeforeUnload({ route });
+	const unloadTimeout = ref<NodeJS.Timeout | null>(null);
 
-	const getUsersForCurrentWorkflow = computed(() => {
-		return usersForWorkflows.value[workflowStore.workflowId] ?? [];
+	addBeforeUnloadHandler(() => {
+		// Notify that workflow is closed straight away
+		notifyWorkflowClosed();
+		if (uiStore.stateIsDirty) {
+			// If user decided to stay on the page we notify that the workflow is opened again
+			unloadTimeout.value = setTimeout(() => notifyWorkflowOpened, 5 * TIME.SECOND);
+		}
 	});
 
+	const collaborators = ref<Collaborator[]>([]);
+
+	const heartbeatTimer = ref<number | null>(null);
+
+	const startHeartbeat = () => {
+		if (heartbeatTimer.value !== null) {
+			clearInterval(heartbeatTimer.value);
+			heartbeatTimer.value = null;
+		}
+		heartbeatTimer.value = window.setInterval(notifyWorkflowOpened, HEARTBEAT_INTERVAL);
+	};
+
+	const stopHeartbeat = () => {
+		if (heartbeatTimer.value !== null) {
+			clearInterval(heartbeatTimer.value);
+		}
+	};
+
+	const pushStoreEventListenerRemovalFn = ref<(() => void) | null>(null);
+
 	function initialize() {
 		if (pushStoreEventListenerRemovalFn.value) {
 			return;
 		}
 
 		pushStoreEventListenerRemovalFn.value = pushStore.addEventListener((event) => {
-			if (event.type === 'activeWorkflowUsersChanged') {
-				const workflowId = event.data.workflowId;
-				usersForWorkflows.value[workflowId] = event.data.activeUsers;
+			if (
+				event.type === 'collaboratorsChanged' &&
+				event.data.workflowId === workflowsStore.workflowId
+			) {
+				collaborators.value = event.data.collaborators;
 			}
 		});
+
+		addBeforeUnloadEventBindings();
+		notifyWorkflowOpened();
+		startHeartbeat();
 	}
 
 	function terminate() {
 		if (typeof pushStoreEventListenerRemovalFn.value === 'function') {
 			pushStoreEventListenerRemovalFn.value();
 			pushStoreEventListenerRemovalFn.value = null;
 		}
-	}
-
-	function workflowUsersUpdated(data: ActiveUsersForWorkflows) {
-		usersForWorkflows.value = data;
-	}
-
-	function functionRemoveCurrentUserFromActiveUsers(workflowId: string) {
-		const workflowUsers = usersForWorkflows.value[workflowId];
-		if (!workflowUsers) {
-			return;
+		notifyWorkflowClosed();
+		stopHeartbeat();
+		pushStore.clearQueue();
```

Comment:
I'm not sure if clearing the queue should be collaborationStores responsibility, as the queue should be private implementation detail of pushStore. Sure it works now, as this is the only one sending messages, but in a general case there could be items from other producers.

Reply:
I think we need to move the queue in this store. but I'd like to do that refactor in another PR.